<a href="https://colab.research.google.com/github/purnawanpp/plants/blob/main/plants.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Protein-Ligand ANTSystem (PLANTS) simulation by Purnawan Pontana Putra**

**Required software installation**

In [1]:
!wget https://github.com/purnawanpp/plants/raw/main/plants
!wget https://github.com/purnawanpp/plants/raw/main/spores
!apt install openbabel
!chmod +x /content/plants
!chmod +x /content/spores
!wget https://github.com/purnawanpp/plants/raw/main/dock_gcolab.txt

--2023-03-09 01:54:13--  https://github.com/purnawanpp/plants/raw/main/plants
Resolving github.com (github.com)... 192.30.255.112
Connecting to github.com (github.com)|192.30.255.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/purnawanpp/plants/main/plants [following]
--2023-03-09 01:54:13--  https://raw.githubusercontent.com/purnawanpp/plants/main/plants
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9536847 (9.1M) [application/octet-stream]
Saving to: ‘plants’

plants              100%[===================>]   9.09M  --.-KB/s    in 0.08s   

2023-03-09 01:54:14 (120 MB/s) - ‘plants’ saved [9536847/9536847]

--2023-03-09 01:54:14--  https://github.com/purnawanpp/plants/raw/main/spores
Reso

**Download Protein PDB ID 3HTB**

In [2]:
!wget https://files.rcsb.org/download/3HTB.pdb

--2023-03-09 01:55:08--  https://files.rcsb.org/download/3HTB.pdb
Resolving files.rcsb.org (files.rcsb.org)... 132.249.210.222
Connecting to files.rcsb.org (files.rcsb.org)|132.249.210.222|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/octet-stream]
Saving to: ‘3HTB.pdb’

3HTB.pdb                [ <=>                ] 161.13K  --.-KB/s    in 0.09s   

2023-03-09 01:55:09 (1.68 MB/s) - ‘3HTB.pdb’ saved [164997]



**Separete Protein and ligand**

This mode will perform a structure recognition and protonation, the output are written in files namely protein.mol2
which contains all protein chains from the input, ligand ligandname_X.mol2 which
contain the ligand molecuels (ligand name is generated from the residue name of
the ligand in the original file, and X is a running number in case there are two
ligands with identical names) and water.mol2 which contains all water molecules
from the input file.

In [3]:
!./spores --mode splitpdb 3HTB.pdb


				 SPORES        

	      Structure PrOtonation and REcognition System

			      version 1.3


author: T. ten Brink

scientific contributors: P. Duchstein, J. Gimmler, O. Korb, M. Zerara,
			  T.E. Exner

contact: tim.tenbrink@uni-konstanz.de


molecule number: 1
64 alternative atomic positions have been removed !
splitting 3HTB
225
protein: 0
ligs: 4
water: 660
metal: 0


**Option Add Hydrogen, charge (gastaiger) and protonation specific pH 7.4**

Information:

⚾!obel: command to run the Obabel program on Google Colab Notebook.

⚾-imol2: option to specify the input file format, i.e. mol2.

⚾/content/protein.mol2: path of the input file to be converted and manipulated.

⚾-omol2: option to specify the output file format, i.e. mol2.

⚾-h: option to add hydrogen to the molecule.

⚾-p 7.4: option to add charge to a molecule with a pH of 7.4.

⚾-b: option to add hydrogen bond information.

⚾--ff "gasteiger": option to specify the calculation force, in this case using the Gasteiger method.

⚾-O rec.mol2: option to specify the name and location of the output file, i.e. rec.mol2




In [6]:
!obabel -imol2 protein.mol2 -omol2 -b --ff -p 7.4 "gasteiger" -O rec.mol2
!obabel -imol2 ligand_JZ4167_0.mol2 -omol2 -b -p 7.4 --ff "gasteiger" -O lig.mol2

*** Open Babel Warning  in ReadMolecule
  Failed to kekulize aromatic bonds in MOL2 file (title is 3HTB_protein)

1 molecule converted
1 molecule converted


**Create and define bindingsite or Gridbox**

In [7]:
!./plants --mode bind lig.mol2 rec.mol2



                  PLANTS
        Protein-Ligand ANT System
               version 1.2


author: Oliver Korb

scientific contributors: T.E. Exner, T. Stuetzle

contact: Oliver.Korb@uni-konstanz.de


run PLANTS: PLANTS --mode screen yourconfigfile



Calculated bindingsite definition (from 1 ligands): 

bindingsite_center 22.7142 -25.2627 -3.283
bindingsite_radius 3.75624



Move data:

* bindingsite_center 22.7142 -25.2627 -3.283
* bindingsite_radius 3.75624

into the file **dock_gcolab.txt**

**Molecular Docking Simulation**

In [8]:
!./plants --mode screen /content/dock_gcolab.txt



                  PLANTS
        Protein-Ligand ANT System
               version 1.2


author: Oliver Korb

scientific contributors: T.E. Exner, T. Stuetzle

contact: Oliver.Korb@uni-konstanz.de


run PLANTS: PLANTS --mode screen yourconfigfile

PLANTS info: CHO hydrogen found 1560 H HIS31
PLANTS info: CHO hydrogen found 1561 H HIS31
virtual screening progress: 1 of 1
current ligand: /content/lig.mol2 (entry 1)
LIGAND DOFs: 9
PROTEIN DOFs: 0
Simplex dimension: 9
Simplex dimension: 9
iterations : 187
starting optimization ...
problem dimension: 9
ATOMS / s: 9.04801e+06
EVAL / s: 904801
optimization finished after 0.81s
best score: -67.44


total virtual screening time: 0.97s
Ligands skipped due to errors: 0


**The ΔG scores or docking results**

After the process is complete, don't forget to click the Google Colab refresh box/file icon. Docking simulation results can be seen in the *results* folder with the file name **bestranking.csv**, see energy TOTAL_SCORE is a delta G value

In [9]:
import pandas as pd
df = pd.read_csv('/content/results/bestranking.csv')
df.head()

LIGAND_ENTRY  TOTAL_SCORE  SCORE_RB_PEN  \
0  3HTB_ligand_2_entry_00001_conf_01     -67.4432      -61.4432   

   SCORE_NORM_HEVATOMS  SCORE_NORM_CRT_HEVATOMS  SCORE_NORM_WEIGHT  \
0             -6.74432                 -31.3044          -0.495229   

   SCORE_NORM_CRT_WEIGHT  SCORE_RB_PEN_NORM_CRT_HEVATOMS  SCORE_NORM_CONTACT  \
0               -13.1087                        -28.5194            -6.74432   

     EVAL      TIME  
0  728868  0.805556

Explanation:

* LIGAND_ENTRY: Name or identification code of the ligand that was docked.
* TOTAL_SCORE: Overall score generated from docking the ligand to the target protein.
* SCORE_RB_PEN: Score for rotamer penalty.
* SCORE_NORM_HEVATOMS: Score for heavy heteroatoms (in this case, normalized value).
* SCORE_NORM_CRT_HEVATOMS: Score for normalized heavy heteroatoms with correction.
* SCORE_NORM_WEIGHT: Score normalized by the weight of the ligand molecule.
* SCORE_NORM_CRT_WEIGHT: Score normalized by correction and weight of the ligand molecule.
* SCORE_RB_PEN_NORM_CRT_HEVATOMS: Score for normalized rotamer and heavy heteroatoms with correction.
* SCORE_NORM_CONTACT: Score for normalized contact.
* EVAL: Evaluation value that measures the quality of docking between the ligand and target protein.
* TIME: Time required to perform the docking.

**OPTIONAL-View the file namely ranking.csv file. It is same as bestranking.csv, but for all ligand poses**

In [10]:
from IPython.display import HTML
import pandas as pd

df = pd.read_csv('/content/results/ranking.csv')

html_table = df.to_html(index=False)
display(HTML(html_table))

**Calculation of RMSD simulation (re-docking)**

In [11]:
!obrms -f lig.mol2 /content/results/3HTB_ligand_2_entry_00001_conf_01.mol2 | sed 's/3HTB_ligand_2:3HTB_ligand_2_entry_00001_conf_01/RE-DOCKING/g'

RMSD RE-DOCKING 0.436417



**Visualization**

In [12]:
!pip install py3Dmol

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [13]:
!obabel /content/rec.mol2 -O recvis.pdb
!obabel /content/results/3HTB_ligand_2_entry_00001_conf_01.mol2 -O ligvis.pdb

1 molecule converted
1 molecule converted


In [14]:
import py3Dmol
v = py3Dmol.view()
v.addModel(open('/content/recvis.pdb').read())
v.setStyle({'cartoon':{},'stick':{'radius':.1}})
v.addModel(open('/content/ligvis.pdb').read())
v.setStyle({'model':1},{'stick':{'colorscheme':'dimgrayCarbon','radius':.125}})
v.addModelsAsFrames(open('/content/ligvis.pdb').read())
v.setStyle({'model':1},{'stick':{'colorscheme':'greenCarbon'}})
v.zoomTo({'model':1})
v.rotate(90)

You appear to be running in JupyterLab (or JavaScript failed to load for some other reason). You need to install the 3dmol extension: 
 jupyter labextension install jupyterlab_3dmol